### PR TITLE
Sanitize header

### LINKED ISSUE #4: Sanitize columns
Using the tap-sftp in combination with a loader such at target-snowflake can trigger errors due to invalid column names. The header in the CSV files can contain characters that are not valid as a snowflake column name.

I personally think this should be handled in the loader, because what constitutes a valid column name varies from target to target. 

@pnadolny13 suggested looking at mappings to solve this, and a regex-based schema mapper looks like a good generic solution. Regardless of what we do with this tap, I think such a mapper would be a useful generic solution when there are issues between taps and targets.

It seems to me that the argument for creating this sanitizer in the tap is mostly pragmatic convenience. Special characters in headers are common in CSV files. Adding an optional keyword in the table config that could solve most issues that arise from column names, would mean a lot of users won't have to add a mapper.

In order to solve our immediate problem, I already created a basic sanitizing option. The main part: https://github.com/radbrt/tap-sftp/blob/a8ce9cc8e9598a64266024d456b30aa6d1b85bb7/tap_sftp/singer_encodings/csv_handler.py#L42

An example config:

```
      tables:
      - table_name: inc_data
        search_prefix: ''
        search_pattern: inc_.*
        key_properties: [key1]
        sanitize_headers: true
        delimiter: ','
        encoding: utf-8
```

### PR DESCRIPTION
**Pull request by radbrt**
Closes https://github.com/MeltanoLabs/tap-sftp/issues/4

Added optional "sanitize_header" config option (bool) to clean file header from special characters.

Replaces non `A-z0-9_` characters with underscore (`_`) and prefixes leading numbers with `x_`.

### CODE REVIEW

File: tap_sftp/singer_encodings/csv_handler.py [hunk 1]
```diff
@@ -2,6 +2,8 @@
 import csv
 import io
 import os
+import singer
```
Comment (pnadolny13) [inline tap_sftp/singer_encodings/csv_handler.py]: @radbrt I dont see any references to this new import, is it needed?

File: tap_sftp/singer_encodings/json_schema.py [hunk 2]
```diff
@@ -38,10 +38,11 @@ def sample_file(conn, table_spec, f, sample_rate, max_records, config):
         file_handle = conn.get_file_handle(f)
 
     # Add file_name to opts and flag infer_compression to support gzipped files
-    opts = {'key_properties': table_spec['key_properties'],
+    opts = {'key_properties': table_spec.get('key_properties', ['_sdc_source_file', '_sdc_source_lineno']),
```
Comment (pnadolny13) [inline tap_sftp/singer_encodings/json_schema.py:41]: @radbrt it looks like this is new behavior of adding default key properties if they arent specified. Is that right? I dont really the implications of this. Is there a chance of users who currently dont have key properties set, getting them added after the PR and upgrading their version? I'm not sure it would matter but want to make sure we dont introduce any breaking changes by accident.

**Review by pnadolny13 — COMMENTED** [review]
@radbrt thanks for the PR! This looks great, I left just a few comments otherwise it looks good to me.

File: tap_sftp/singer_encodings/csv_handler.py (hunk 1 above)
Comment (radbrt) [reply tap_sftp/singer_encodings/csv_handler.py]: True, I have no clue how that got there. Removing.

File: tap_sftp/singer_encodings/json_schema.py (hunk 2 above)
Comment (radbrt) [reply tap_sftp/singer_encodings/json_schema.py:41]: Agreed, My thought was that I would provide a default in case nothing was specified, but this would override if the tap was configured with an empty list. And that would be a breaking change, which we don't want. Removing.

Comment (pnadolny13) [reply tap_sftp/singer_encodings/json_schema.py:41]: @radbrt sorry for the slow response. If you "@" me with new comments I'll get a notification so I dont miss anything :smile:!

>My thought was that I would provide a default in case nothing was specified, but this would override if the tap was configured with an empty list. And that would be a breaking change

I actually think it would respect an empty list because the `get('key_properties'` so only null key properties would have a change. The more I think about it I think I'm actually in favor of making that the default, especially because this is a new repo and we dont have any releases yet so most likely users are using the bonobos variant still or are pinning commit hashes.

What do you think? If you agree, would you mind adding a note in the readme explaining it and also add `sanitize_header` as one of the options available for the tables entries.

FWIW I think this is doing the same thing so whatever we decide we should make sure its consistent across the two https://github.com/MeltanoLabs/tap-sftp/pull/11/files#diff-cb5a8465169f4b210c9601bbc6ba2988f90598e197fd637dc378f0ca20e74cf5L63.

Comment (radbrt) [reply tap_sftp/singer_encodings/json_schema.py:41]: Thanks @pnadolny13 

I tried it out just to be sure, and the `get` method does seem to respect an empty list. So basically:

- Existing installations must have configured a key, otherwise it would have failed. So there are no functioning existing installations without a `key_properties` config key set.
- If we add a default in the `get` property, it will respect anything configured (including empty lists), except if no `key_properties` are set, in which case it wouldn't be a functioning installation. So adding a default would not impact any currently functioning installation.

It does seem fairly safe to add a default here.

I keep forgetting that the parameters are passed through in two different files, so thanks for reminding me I need to change it two places.

In summary, it seems we agree about the following two changes:
- revert to a default key_properties, and follow through by marking it as optional
- make sure we pass those defaults on in all places that propagates config
- document the sanitize_header config key.

As a final question: I did not put much thought into the `sanitize_header` key name, and I don't know if it is the best name. I have toyed around with `clean_colnames` as an alternative, but that might be more ambigious. Thoughts?

Comment (pnadolny13) [reply tap_sftp/singer_encodings/json_schema.py:41]: @radbrt yes that summary sounds correct to me! Personally I'm fine with `sanitize_header` for naming. Once we have an entry in the README and MeltanoHub it should be clear what it does.

Comment (radbrt) [reply tap_sftp/singer_encodings/json_schema.py:41]: @pnadolny13 I think the newest updates should do it. Tests passed, have tested manually with some different configurations to make sure an empty key_property key won't be overwritten, but that when the config key isn't specified the default kicks in.

File: README.md [hunk 3]
```diff
@@ -55,7 +56,25 @@ $ pip install .
...
+   - `tables`: An array of tables to load. See table configuration below
+   - `start_date`: Earliest file date to synchronize
+   - either `password` or `private_key_file`: Authentication to the SFTP Server
+
+
+   The following table configuration fields are required:
+   - `table_name`: The name that should be given to the table (stream)
+   - `search_prefix`: Folder where the files are located
+   - `search_pattern`: Regex pattern to match the file names
+   - `delimiter`: CSV file delimiter,
```
Comment (pnadolny13) [inline README.md]:
```suggestion
   - `delimiter`:  A one-character string delimiter used to separate fields. Default, is `,`.
```

File: README.md [hunk 4]
```diff
@@ -55,7 +56,25 @@ $ pip install .
...
+
+   The following table configuration fields are optional:
+   - `key_properties`: Array containing the unique keys of the table. Defaults to `['_sdc_source_file', '_sdc_source_lineno']`, representing the file name and line number. Specify an emtpy array (`[]`) to load all new files without a replication key
+   - `encoding`: File encoding, defaults to `utf-8`
+   - `sanitize_header`: Boolean, specifies wether to clean up header names so that they are more likely to be accepted by a target SQL database.
```
Comment (pnadolny13) [inline README.md]:
```suggestion
   - `sanitize_header`: Boolean, specifies whether to clean up header names so that they are more likely to be accepted by a target SQL database.
```

**Review by pnadolny13 — APPROVED** [review]
@radbrt I left a few suggestions for readme typos but other than that this looks good to me! Thanks for getting it across the finish line.

Feel free to merge it in once you're comfortable (let me know if you dont have permissions but you should).

File: README.md (hunk 3 above)
Comment (radbrt) [reply README.md]: I tried making delimiter optional, but it would have required more changes than I had thought and so I considered it out of scope here. So although comma is used as default several places in the code, it isn't optional and so it kinda doesn't have a default. But we'll leave that for another day.
